SNMP Set Node

mscbpi opened the pull request:
@bimalyn-IBM 
Follow-up of #311 based on updated master branch (sorry I think there was a better way to reset the initial fork I guess).

bimalyn-IBM commented:
You should not have needed to open a new PR. Check out this [guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) to learn how to add a remote from fork and this [guide](https://help.github.com/articles/syncing-a-fork/) to learn how to pull the changes from upstream.

I don't know if it really matters at this point which PR you close.

mscbpi commented:
Thanks and sorry, still learning. I closed the previous one since it's referring to a deleted rep.

bimalyn-IBM commented:
You need to pull from upstream again. The `packages.json` file has been modified to have the correct version of `net-snmp`.

Another option you have, to pull the code from upstream, is to do the following:
```
git stash
git pull upstream master
git stash apply
```

bimalyn-IBM commented:
Have you tested your code? I'm having difficulty setting a value.

The value I try to set (this is after the type has been modified) is `[ { oid: '1.3.6.1.2.1.1.5.0', type: 4, value: 'Test' } ]` and the error I am receiving is `RequestFailedError: BadValue: 1.3.6.1.2.1.1.5.0`

mscbpi commented:
I am actually using it yes. Shouldn't you write type: 'OctetString' ?

bimalyn-IBM commented:
I did, that was the output from here:
```javascript
...
for (var i = 0; i < varbinds.length; i++) {
    varbinds[i].type = objectTypes[varbinds[i].type];
}
console.log(varbinds);
...
```

mscbpi commented:
Ok, I had no issues with integer ==> type 2, but I am going to test this one as well. Thanks for the feedback.

bimalyn-IBM commented:
Oh, I'm sorry. I just read the mib again and it turns out that was a read-only oid. I don't have a mib that I can write a string to. Just one that allows an integer.

mscbpi commented:
👍I will find one and test, it's worth it.

bimalyn-IBM commented:
I noticed that you also changed the names of the objectTypes for example `Integer` vs `Integer32`.

bimalyn-IBM commented on io/snmp/snmp.js:
```diff
@@ -52,14 +52,72 @@ module.exports = function(RED) {
+            var varbinds = msg.varbinds;
+
+            if (varbinds) {
+                for (var i = 0; i < varbinds.length; i++) {
```
```javascript
for (var i = 0; i < varbinds.length; i++) {
    varbinds[i].type = snmp.ObjectType[varbinds[i].type];
}
```

mscbpi replied:
I actually copy pasted what you put below:

```
var ObjectType = {
	1: "Boolean",
	2: "Integer",
	4: "OctetString",
	5: "Null",
	6: "OID",
	64: "IpAddress",
	65: "Counter",
	66: "Gauge",
	67: "TimeTicks",
	68: "Opaque",
	70: "Counter64",
	128: "NoSuchObject",
	129: "NoSuchInstance",
	130: "EndOfMibView"
};
```

But I needed it the other way around (`"String": Integer`). I'll search for a better code to use this ( `eval()` is not right either, is it ?).

bimalyn-IBM replied:
ObjectType also has the keys/value pairs switched. You don't need an `eval()`. You can use the line as I wrote it in the above comment.

mscbpi replied:
My bad, I read too fast...

dceejay commented:
a) what does the standard say ? or if in doubt handle both ?

bimalyn-IBM commented:
```javascript 
function _expandConstantObject (object) {
	var keys = [];
	for (var key in object)
		keys.push (key);
	for (var i = 0; i < keys.length; i++)
		object[object[keys[i]]] = parseInt (keys[i]);
}

var ObjectType = {
	1: "Boolean",
	2: "Integer",
	4: "OctetString",
	5: "Null",
	6: "OID",
	64: "IpAddress",
	65: "Counter",
	66: "Gauge",
	67: "TimeTicks",
	68: "Opaque",
	70: "Counter64",
	128: "NoSuchObject",
	129: "NoSuchInstance",
	130: "EndOfMibView"
};

_expandConstantObject (ObjectType);

ObjectType.Integer32 = ObjectType.Integer;
ObjectType.Counter32 = ObjectType.Counter;
ObjectType.Gauge32 = ObjectType.Gauge;
ObjectType.Unsigned32 = ObjectType.Gauge32;

```

So both are supported if we do the lookups directly in the `snmp.ObjectType`, but not in the hard-coded values of `objectTypes`

mscbpi commented on io/snmp/snmp.html:
```diff
@@ -57,6 +57,58 @@
+            <option value="2c">v2c</option>
+        </select>
+    </div>
+    <div class="form-row">
```
How would you do it ? via JSON object in the node _(raw) ? Ideal would have been to be able to set mutiple varbinds in the node similar to msg.varbinds.

bimalyn-IBM replied:
I'm not sure if that would work or if you'd need to use a [configuration node](https://nodered.org/docs/creating-nodes/config-nodes). @dceejay What would you suggest?

mscbpi replied:
A configuration node would be to set up connection for all type of snmp node or just snmp set ? Would you set content/variables (in that case varbinds) in the configuration node or just manage connections there (host/community/version) ?

bimalyn-IBM replied:
Maybe looking at the approach used in the switch node would help (specifically the rules section) [html](https://github.com/node-red/node-red/blob/master/nodes/core/logic/10-switch.html) and [js](https://github.com/node-red/node-red/blob/master/nodes/core/logic/10-switch.js).

mscbpi replied:
This could be applied to all node (multiple oids seperate by comma ATM). If we get there, snmp-set would add the type and value field, but IMO we have to have, when possible, a similar experience between nodes. I agree that at the moment there is nothing for varbinds in the node.

bimalyn-IBM replied:
`get` is the only other node that takes multiple oids, the others will only accept a single oid.

Modifying `get` would likely break old deployments of this node, when upgrading. I don't know if there would be a clean way to do this.

mscbpi commented on io/snmp/snmp.js:
```diff
@@ -52,14 +52,72 @@ module.exports = function(RED) {
+                });
+            }
+            else {
+                node.warn("No oid(s) to set");
```
Right.

mscbpi commented:
@bimalyn-IBM @dceejay I have updated with some changes. node.varbinds is currently possible as stirng (array of JSON object). I corrected priorities between properties set in node and message (node has priority, same as the other nodes). Corrected documentation.

bimalyn-IBM commented on io/snmp/snmp.html:
```diff
@@ -35,33 +34,93 @@
+    <p>Simple snmp Set node. Trigger by any input</p>
+    <p><code>msg.host</code> may contain the host.</p>
+    <p><code>msg.community</code> may contain the community.</p>
+    <p><code>msg.varbinds</code> may contain varbinds as an Array [ { "oid": "1.3.6.1.2.1.1.5.0", "type": "OctetString", "value":
```
Can you wrap the array in a `<code>` block?

mscbpi commented:
@bimalyn-IBM corrected code, thanks. I would have said that a wrong varbinds would have trigger an error in session.set().

bimalyn-IBM commented on io/snmp/snmp.js:
```diff
@@ -65,7 +65,18 @@ module.exports = function (RED) {
+                    varbinds=JSON.parse(node.varbinds);
+                }
+                catch (e) {
+                    node.error("Error parsing varbinds ");
```
`node.error("Error parsing varbinds ");` won't be raised to the catch block without adding the `msg` parameter: `node.error("Error parsing varbinds", msg);`

bimalyn-IBM commented:
It's probably true that an error would have been raised from an invalid varbinds object. You would have had to test that. If that is the case I don't care if you catch it explicitly.

mscbpi commented:
That is the case, removing the try, catch.
![image](https://user-images.githubusercontent.com/28528210/27343161-5ee309b6-55e2-11e7-8c94-5c60e87aecfe.png)

bimalyn-IBM commented:
Excellent. Let me pull all the code down and see how it all looks. I assume you've tested the set node? I don't really have any devices that I can do a set for a string, int, etc.

bimalyn-IBM commented on io/snmp/snmp.html:
```diff
@@ -35,33 +34,91 @@
+    <p>Simple snmp Set node. Trigger by any input</p>
+    <p><code>msg.host</code> may contain the host.</p>
+    <p><code>msg.community</code> may contain the community.</p>
+    <p><code>msg.varbinds</code> may contain varbinds as an array of json objects containing multiple oids, types and values.<code>[ { "oid": "1.3.6.1.2.1.1.5.0","type": "OctetString", "value": "host1"} , { "oid": ...} ]</code></p>
```
Maybe we should break this line up like this:

```javascript
     <p><code>msg.varbinds</code> may contain varbinds as an array of json objects containing multiple oids, types and values.
        <pre>[
    {
        "oid":   "1.3.6.1.2.1.1.5.0",
        "type":  "OctetString",
        "value": "host1"
    },
    {   "oid":   ... }
]</pre>
    </p>
```

bimalyn-IBM replied:
![snmp-set-varbinds](https://user-images.githubusercontent.com/26979538/27355180-ada20b82-55cf-11e7-8545-3fe416806c95.jpg)

mscbpi replied:
ok, to make it all at once, what do you think in placeholder area ? I can do line breaks (&10;), but not sure about tabs.

bimalyn-IBM replied:
I wouldn't worry about the placeholder, most browsers don't allow line breaks in the placeholder. In the [specification](https://www.w3.org/TR/html5/forms.html#the-placeholder-attribute) for placeholders it says: 

> The attribute [placeholder], if specified, must have a value that contains **no** "LF" (U+000A) [&amp;#10;] or "CR" (U+000D) [&amp;#13;] characters.

Tabs work fine (&amp;#9;), but without line breaks they are not very helpful.

bimalyn-IBM commented:
I think that this looks like it's ready to be pulled. @mscbpi Do you have any more changes?

mscbpi commented:
I'm all good !

dceejay commented:
Thanks @mscbpi @bimalyn-IBM

bimalyn-IBM commented:
You're welcome.

mscbpi commented:
you're welcome, and thanks @bimalyn-IBM  for the tips !

bimalyn-IBM commented:
No problem. Glad to have been of help!

qlartpg commented:
In the snmp set node, if I assign the varbinds to:
`[{"oid":"1.3.6.1.4.1.13742.4.1.2.2.1.3.1","type":"Integer","value":1}]`
The set var node works great. However, if I use a function to assign varbinds:
```
msg.varbinds = [{

    "oid": "1.3.6.1.4.1.13742.4.1.2.2.1.3.1",

    "type": "Integer",

    "value": 1 

}];
```
I would get a ResquestTimedOutError. Can someone please let me know what I am doing wrong?
Thanks!

mscbpi commented:
Anything in the logs apart from timeout ? May you remove the line breaks in your varbinds? Can you try to test to JSON.parse your varbinds within nodered ? Either using JSON.parse() or JSON  nodered node.

qlartpg commented:
Sorry I'm new to node red and Javascript. Here is an example of my flow:

`[{"id":"d996b8f1.7c0138","type":"ui_switch","z":"baf2641a.f34178","name":"Swith 01","label":"","group":"18815ca9.8ffb93","order":3,"width":"1","height":"3","passthru":false,"decouple":"false","topic":"","style":"","onvalue":"1","onvalueType":"num","onicon":"","oncolor":"","offvalue":"-1","offvalueType":"num","officon":"","offcolor":"","x":671.0000686645508,"y":67.0000057220459,"wires":[["e22030d3.8fd01","f1f788a.f60d778"]]},{"id":"f1f788a.f60d778","type":"switch","z":"baf2641a.f34178","name":"","property":"payload","propertyType":"msg","rules":[{"t":"gt","v":"0","vt":"num"},{"t":"lt","v":"0","vt":"num"}],"checkall":"true","outputs":2,"x":917.000093460083,"y":105.00000762939453,"wires":[["c703496.7357bb8"],["b04af061.d266"]]},{"id":"c703496.7357bb8","type":"snmp set","z":"baf2641a.f34178","host":"192.168.41.131","community":"raritan_private","version":"2c","varbinds":"[{\"oid\":\"1.3.6.1.4.1.13742.4.1.2.2.1.3.1\",\"type\":\"Integer\",\"value\":1}]","name":"TurnOutletOn","x":1111.000099182129,"y":68.0000114440918,"wires":[]},{"id":"b04af061.d266","type":"snmp set","z":"baf2641a.f34178","host":"192.168.41.131","community":"raritan_private","version":"2c","varbinds":"[{\"oid\":\"1.3.6.1.4.1.13742.4.1.2.2.1.3.1\",\"type\":\"Integer\",\"value\":0}]","name":"TurnOutletOFF","x":1320.9999542236328,"y":85.00001430511475,"wires":[]},{"id":"88fa47cd.8322c8","type":"debug","z":"baf2641a.f34178","name":"","active":true,"console":"false","complete":"true","x":1282.0000610351562,"y":264.00005054473877,"wires":[]},{"id":"544846cf.234a88","type":"function","z":"baf2641a.f34178","name":"","func":"outlet = msg.payload\nif (outlet <0) {outlet = -outlet;}\n\nvar oid\noid = \"1.3.6.1.4.1.13742.4.1.2.2.1.3.\".concat(outlet);\n\nif (msg.payload > 0){\n    //turn on\n    value = 1;\n} else {\n    //turn off\n    value = 0;\n}\n\n//example\n//[{\"oid\":\"1.3.6.1.4.1.13742.4.1.2.2.1.3.1\",\"type\":\"Integer\",\"value\":1}]\n\nmsg.varbinds = [{\n    \"oid\":\"1.3.6.1.4.1.13742.4.1.2.2.1.3.1\",\n    \"type\":\"Integer\",\n    \"value\":value\n    \n}];\n    \n//msg.varbinds = [{\"oid\":\"1.3.6.1.4.1.13742.4.1.2.2.1.3.2\",\"type\":\"Integer\",\"value\":0}]   \n    \n//msg.payload = msg.varbinds\n\nreturn msg","outputs":1,"noerr":0,"x":1130.0002098083496,"y":207.00002670288086,"wires":[["88fa47cd.8322c8","f15daabc.2ff148"]]},{"id":"ade8c737.2ea768","type":"inject","z":"baf2641a.f34178","name":"","topic":"","payload":"-1","payloadType":"num","repeat":"","crontab":"","once":false,"x":664.3907890319824,"y":220.12894821166992,"wires":[["e22030d3.8fd01","f1f788a.f60d778"]]},{"id":"f15daabc.2ff148","type":"snmp set","z":"baf2641a.f34178","host":"","community":"raritan_private","version":"1","varbinds":"","name":"TurnOutletOn/Off","x":1319.7656631469727,"y":205.75002479553223,"wires":[]},{"id":"e22030d3.8fd01","type":"function","z":"baf2641a.f34178","name":"192.168.41.131","func":"//msg.host = global.get(\"GlobalHostIP\")\n\nmsg.Host = \"192.168.41.131\"\nreturn msg;","outputs":1,"noerr":0,"x":947.765754699707,"y":205.75000953674316,"wires":[["544846cf.234a88"]]},{"id":"fa3d9068.61c0b","type":"inject","z":"baf2641a.f34178","name":"","topic":"","payload":"1","payloadType":"num","repeat":"","crontab":"","once":false,"x":665.7656631469727,"y":151.7500295639038,"wires":[["e22030d3.8fd01","f1f788a.f60d778"]]},{"id":"30b3fe6d.9d9272","type":"comment","z":"baf2641a.f34178","name":"This bottom does not work","info":"This bottom one is essentially similar to the top.\nHowever, It does not work","x":1144.895751953125,"y":157.88888549804688,"wires":[]},{"id":"605f16a5.987bd8","type":"comment","z":"baf2641a.f34178","name":"Top works","info":"setting varbines in snmp node set works","x":876.8958129882812,"y":39.888885498046875,"wires":[]},{"id":"18815ca9.8ffb93","type":"ui_group","z":"","name":"Outlet 01","tab":"a9db8836.b5dc58","order":1,"disp":false,"width":"6"},{"id":"a9db8836.b5dc58","type":"ui_tab","z":"","name":"PDU130","icon":"dashboard","order":2}]`

mscbpi commented:
Have you tried removing line breaks: 
`msg.varbinds = [{"oid":"1.3.6.1.4.1.13742.4.1.2.2.1.3.1","type":"Integer","value":value}];`

mscbpi commented:
Also:
1.
`msg.Host = "192.168.41.131"` has to be corrected to
`msg.host = "192.168.41.131"`

2. in your working example, the node is set to version v2c while it is v1 in the non-working node.

qlartpg commented:
OMG!!! Arrrrghhhh!! The whole day of trying to figured it out and it is "msg.host" that is causing all of the problem!! Thank you! Thank you!

dweston commented:
> No problem. Glad to have been of help!

Hi @bimalyn-IBM , I'm attempting to use this node to enable/disable a PoE on a specific port on a GS728TP switch, but running into a problem with the node returning a "TypeError: argument must be a Number" error.

**Background:**
The following example command works fine using `Net-SNMP` on Linux.

`$ snmpset -v 2c -c [community-string] [device-IP-address] POWER-ETHERNET-MIB::pethPsePortAdminEnable.1.24 = 2`

Using _Wireshark_ to view the `snmp` packets being exchanged between Linux `Net-SNMP` MIB controller and the switch appears to confirm that the data type for port status Value should be `Integer32`. Note its the same result whether I use snmp v1 or v2c.
![Screen Shot 2019-12-08 at 22 00 50](https://user-images.githubusercontent.com/631241/71334481-8977bc80-2592-11ea-88f0-9b9f57fa3e27.jpg)

But using that data type in the node set-up generates the TypeError mentioned above.
![Screen Shot 2019-12-08 at 22 03 56](https://user-images.githubusercontent.com/631241/71334603-f2f7cb00-2592-11ea-828e-5fc13d1e38a8.jpg)
![Screen Shot 2019-12-08 at 22 04 31](https://user-images.githubusercontent.com/631241/71334616-fdb26000-2592-11ea-8187-e8f1364bd85d.jpg)
 
I've tried using TypeData "Integer", "Integer32" and even "OctetString", though the latter gives a different error message:
`"RequestFailedError: WrongType: 1.3.6.1.2.1.105.1.1.1.3.1.22"`

Would you have a suggestion as to how I might troubleshoot this issue?

dweston commented:
This is the test flow I'm using:
`[
    {
        "id": "e248012c.14fdb",
        "type": "snmp",
        "z": "93143d3a.2ea878",
        "host": "10.0.1.170",
        "community": "test",
        "version": "2c",
        "oids": "1.3.6.1.2.1.105.1.1.1.3.1.24",
        "timeout": 5,
        "name": "Get PoE State PoE - port 24",
        "x": 380,
        "y": 500,
        "wires": [
            [
                "c763f70a.d3baf"
            ]
        ]
    },
    {
        "id": "c763f70a.d3baf",
        "type": "debug",
        "z": "93143d3a.2ea878",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": false,
        "tostatus": false,
        "complete": "false",
        "x": 640,
        "y": 500,
        "wires": []
    },
    {
        "id": "8e8b2f45.13f86",
        "type": "inject",
        "z": "93143d3a.2ea878",
        "name": "",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "x": 120,
        "y": 500,
        "wires": [
            [
                "e248012c.14fdb"
            ]
        ]
    },
    {
        "id": "fa3adb5c.bfc84",
        "type": "snmp set",
        "z": "93143d3a.2ea878",
        "host": "10.0.1.170",
        "community": "test",
        "version": "2c",
        "varbinds": "[\n    {\n        \"oid\":   \"1.3.6.1.2.1.105.1.1.1.3.1.24\",\n        \"type\":  \"Integer32\",\n        \"value\": \"2\"\n    }\n]",
        "timeout": 5,
        "name": "PoE Disable Port 24",
        "x": 360,
        "y": 560,
        "wires": []
    },
    {
        "id": "5aabbed6.db3038",
        "type": "inject",
        "z": "93143d3a.2ea878",
        "name": "",
        "topic": "",
        "payload": "",
        "payloadType": "date",
        "repeat": "",
        "crontab": "",
        "once": false,
        "onceDelay": 0.1,
        "x": 120,
        "y": 560,
        "wires": [
            [
                "fa3adb5c.bfc84"
            ]
        ]
    }
]`

hardillb commented:
@dweston take the quotes out from round the 2

[
  {
    "oid": "1.3.6.1.2.1.105.1.1.1.3.1.2",
   "type": "Integet32",
   "value": 2
  }
]

dweston commented:
You are a gem @hardillb . Thank you so much. Shows me just how much I still have to get my head around json syntax.

dceejay commented:
I have added a comment to the node version 0.0.25 to hopefully make that more clear.
